### PR TITLE
Separate electron imports between main and renderer, use IPC to replace most renderer-side electron imports

### PR DESCRIPTION
**Pull request by joeyballentine**
electron provides (and recommends using) these segregated imports (electron/renderer, electron/main, electron/common) in order to ensure you aren't accidentally importing something in the wrong place.

I discovered a few places where we were importing main-process electron code in the renderer and have fixed it, along with modifying (almost) all the electron imports to be the helper ones. (I left one import the same because if I didn't then it thinks electron doesn't exist. Try it yourself, its really strange)

I'm opening this as a draft, because there's still somewhere that is importing main process code where it shouldn't, and it is causing chaiNNer to crash on startup.

### CODE REVIEW

**Comment by joeyballentine** [comment]
So, I learned a couple things:
1) the electron webpack template is bad and you apparently need to externalize the modules for them to be imported properly. This took the most time to debug
2) the namespace imports _do_ prevent you from using them in the renderer, since you're supposed to only expose stuff via the preload script. This is unlike the main electron import that doesn't care. This means that i had to leave a few as regular electron imports in the meantime, until we swap over to exposing IPC via the preload script
3) Because of the above, i merged a separate branch into this one where i had replaced a lot of renderer-side imports with IPC calls. Theoretically these could be replaced by instead exposing the clipboard api through the preload script, but for now it was just easier to send that stuff through IPC and have IPC be the only thing still being imported the incorrect way. Then, we can move IPC over to the correct way, then replace some of the IPC with non-IPC in the future.

File: src/common/safeIpc.ts [hunk 1]
```diff
@@ -1,13 +1,4 @@
...
-    BrowserWindow,
-    IpcMainEvent,
-    IpcMainInvokeEvent,
-    IpcRendererEvent,
-    MessagePortMain,
-    WebContents,
-    ipcMain as unsafeIpcMain,
-    ipcRenderer as unsafeIpcRenderer,
-} from 'electron';
+import { type FileFilter, type OpenDialogReturnValue } from 'electron'; // TODO: replace with electron/common
```
Comment (RunDevelopment) [inline src/common/safeIpc.ts:1]: TODO

File: src/renderer/components/inputs/DirectoryInput.tsx [hunk 2]
```diff
@@ -90,7 +93,9 @@ export const DirectoryInput = memo(
                     isDisabled={!displayDirectory}
                     onClick={() => {
                         if (displayDirectory) {
-                            clipboard.writeText(displayDirectory);
+                            ipcRenderer
+                                .invoke('clipboard-writeText', displayDirectory)
+                                .catch(log.error);
```
Comment (RunDevelopment) [inline src/renderer/components/inputs/DirectoryInput.tsx:98]: Why were we using electron clipboard in the first place? There's a [browser API for this](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText). Wouldn't that work too?

File: src/renderer/contexts/GlobalNodeState.tsx [hunk 3]
```diff
@@ -28,7 +28,7 @@ import { log } from '../../common/log';
 import { getEffectivelyDisabledNodes } from '../../common/nodes/disabled';
 import { ChainLineage } from '../../common/nodes/lineage';
 import { TypeState } from '../../common/nodes/TypeState';
-import { ipcRenderer } from '../../common/safeIpc';
+
```
Comment (RunDevelopment) [inline src/renderer/contexts/GlobalNodeState.tsx:31]: weird empty line
```suggestion
```

File: src/renderer/hooks/useIpcRendererListener.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
-import { IpcRendererEvent } from 'electron';
+import { IpcRendererEvent } from 'electron'; // TODO: replace with electron/renderer
```
Comment (RunDevelopment) [inline src/renderer/hooks/useIpcRendererListener.ts:1]: TODO

File: src/renderer/safeIpc.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { IpcRendererEvent, ipcRenderer as unsafeIpcRenderer } from 'electron'; // TODO: replace with electron/renderer
```
Comment (RunDevelopment) [inline src/renderer/safeIpc.ts:1]: TODO

File: src/common/safeIpc.ts (hunk 1 above)
Comment (joeyballentine) [reply src/common/safeIpc.ts:1]: That's not a now todo, that's a "we can't do that til we do more stuff in the future" todo. As in, if I were to do that now this wouldn't compile.

File: src/renderer/components/inputs/DirectoryInput.tsx (hunk 2 above)
Comment (joeyballentine) [reply src/renderer/components/inputs/DirectoryInput.tsx:98]: Probably yeah, I can check that

File: src/renderer/contexts/GlobalNodeState.tsx (hunk 3 above)
Comment (joeyballentine) [reply src/renderer/contexts/GlobalNodeState.tsx:31]: Sorry I thought I removed all these. Was from bad merge resolutions

File: src/renderer/hooks/useIpcRendererListener.ts (hunk 4 above)
Comment (joeyballentine) [reply src/renderer/hooks/useIpcRendererListener.ts:1]: Ditto above

File: src/renderer/safeIpc.ts (hunk 5 above)
Comment (joeyballentine) [reply src/renderer/safeIpc.ts:1]: Ditto above

File: src/common/safeIpc.ts (hunk 1 above)
Comment (joeyballentine) [reply src/common/safeIpc.ts:1]: I also explained this in point 2 of my comment. For reference:

> This means that i had to leave a few as regular electron imports in the meantime, until we swap over to exposing IPC via the preload script

File: src/renderer/components/inputs/DirectoryInput.tsx (hunk 2 above)
Comment (joeyballentine) [reply src/renderer/components/inputs/DirectoryInput.tsx]: Ok, I replaced all the read/write text ones. Can't replace the others as the native clipboard doesn't support as many formats as electron's does.

**Comment by joeyballentine** [comment]
Ok, back to a draft. i broke chain copy and paste from this

**Comment by joeyballentine** [comment]
I think i've tested all the different ways to copy and paste, and everything looks good now.

**Review by RunDevelopment — APPROVED** [review]

**Pull request merged**